QMOLEDEV/cmake-2.6.0/Docs/ccmake.1
.TH ccmake 1 "December 19, 2011" "ccmake 2.6-patch 0"
.SH NAME
.PP
  ccmake - Curses Interface for CMake.
.SH USAGE
.PP
  ccmake <path-to-source>
  ccmake <path-to-existing-build>
.SH DESCRIPTION
.PP
The "ccmake" executable is the CMake curses interface.  Project configuration settings may be specified interactively through this GUI.  Brief instructions are provided at the bottom of the terminal when the program is running.

.PP
CMake is a cross-platform build system generator.  Projects specify their build process with platform-independent CMake listfiles included in each directory of a source tree with the name CMakeLists.txt. Users build a project by using CMake to generate a build system for a native tool on their platform.

.SH OPTIONS
.TP
.B -C <initial-cache>
Pre-load a script to populate the cache.

When cmake is first run in an empty build tree, it creates a CMakeCache.txt file and populates it with customizable settings for the project.  This option may be used to specify a file from which to load cache entries before the first pass through the project's cmake listfiles.  The loaded entries take priority over the project's default values.  The given file should be a CMake script containing SET commands that use the CACHE option, not a cache-format file.

.TP
.B -D <var>:<type>=<value>
Create a cmake cache entry.

When cmake is first run in an empty build tree, it creates a CMakeCache.txt file and populates it with customizable settings for the project.  This option may be used to specify a setting that takes priority over the project's default value.  The option may be repeated for as many cache entries as desired.

.TP
.B -U <globbing_expr>
Remove matching entries from CMake cache.

This option may be used to remove one or more variables from the CMakeCache.txt file, globbing expressions using * and ? are supported. The option may be repeated for as many cache entries as desired.


Use with care, you can make your CMakeCache.txt non-working.

.TP
.B -G <generator-name>
Specify a makefile generator.

CMake may support multiple native build systems on certain platforms.  A makefile generator is responsible for generating a particular build system.  Possible generator names are specified in the Generators section.

.TP
.B -Wno-dev
Suppress developer warnings.

Suppress warnings that are meant for the author of the CMakeLists.txt files.

.TP
.B -Wdev
Enable developer warnings.

Enable warnings that are meant for the author of the CMakeLists.txt files.

.SH GENERATORS
.TP
.B Unix Makefiles
Generates standard UNIX makefiles.

A hierarchy of UNIX makefiles is generated into the build tree.  Any standard UNIX-style make program can build the project through the default make target.  A "make install" target is also provided.

.TP
.B CodeBlocks - Unix Makefiles
Generates CodeBlocks project files.

Project files for CodeBlocks will be created in the top directory and in every subdirectory which features a CMakeLists.txt file containing a PROJECT() call. Additionally a hierarchy of makefiles is generated into the build tree.  The appropriate make program can build the project through the default make target.  A "make install" target is also provided.

.TP
.B Eclipse CDT4 - Unix Makefiles
Generates Eclipse CDT 4.0 project files.

Project files for Eclipse will be created in the top directory and will have a linked resource to every subdirectory which features a CMakeLists.txt file containing a PROJECT() call.Additionally a hierarchy of makefiles is generated into the build tree. The appropriate make program can build the project through the default make target. A "make install" target is also provided.

.TP
.B KDevelop3
Generates KDevelop 3 project files.

Project files for KDevelop 3 will be created in the top directory and in every subdirectory which features a CMakeLists.txt file containing a PROJECT() call. If you change the settings using KDevelop cmake will try its best to keep your changes when regenerating the project files. Additionally a hierarchy of UNIX makefiles is generated into the build tree.  Any standard UNIX-style make program can build the project through the default make target.  A "make install" target is also provided.

.TP
.B KDevelop3 - Unix Makefiles
Generates KDevelop 3 project files.

Project files for KDevelop 3 will be created in the top directory and in every subdirectory which features a CMakeLists.txt file containing a PROJECT() call. If you change the settings using KDevelop cmake will try its best to keep your changes when regenerating the project files. Additionally a hierarchy of UNIX makefiles is generated into the build tree.  Any standard UNIX-style make program can build the project through the default make target.  A "make install" target is also provided.

.SH PROPERTIES
.PP
  CMake Properties - Properties supported by CMake, the Cross-Platform Makefile Generator.
.PP
This is the documentation for the properties supported by CMake. Properties can have different scopes. They can either be assigned to a source file, a directory, a target or globally to CMake. By modifying the values of properties the behaviour of the build system can be customized.

.SH COMMAND
.TP
.B add_custom_command
Add a custom build rule to the generated build system.

There are two main signatures for add_custom_command The first signature is for adding a custom command to produce an output.


  add_custom_command(OUTPUT output1 [output2 ...]
                     COMMAND command1 [ARGS] [args1...]
                     [COMMAND command2 [ARGS] [args2...] ...]
                     [MAIN_DEPENDENCY depend]
                     [DEPENDS [depends...]]
                     [IMPLICIT_DEPENDS <lang1> depend1 ...]
                     [WORKING_DIRECTORY dir]
                     [COMMENT comment] [VERBATIM] [APPEND])

This defines a new command that can be executed during the build process. The outputs named should be listed as source files in the target for which they are to be generated. If an output name is a relative path it will be interpreted relative to the build tree directory corresponding to the current source directory. Note that MAIN_DEPENDENCY is completely optional and is used as a suggestion to visual studio about where to hang the custom command. In makefile terms this creates a new target in the following form:


  OUTPUT: MAIN_DEPENDENCY DEPENDS
          COMMAND

If more than one command is specified they will be executed in order. The optional ARGS argument is for backward compatibility and will be ignored.


The second signature adds a custom command to a target such as a library or executable. This is useful for performing an operation before or after building the target. The command becomes part of the target and will only execute when the target itself is built.  If the target is already built, the command will not execute.


  add_custom_command(TARGET target
                     PRE_BUILD | PRE_LINK | POST_BUILD
                     COMMAND command1 [ARGS] [args1...]
                     [COMMAND command2 [ARGS] [args2...] ...]
                     [WORKING_DIRECTORY dir]
                     [COMMENT comment] [VERBATIM])

This defines a new command that will be associated with building the specified target. When the command will happen is determined by which of the following is specified:


  PRE_BUILD - run before all other dependencies
  PRE_LINK - run after other dependencies
  POST_BUILD - run after the target has been built

Note that the PRE_BUILD option is only supported on Visual Studio 7 or later. For all other generators PRE_BUILD will be treated as PRE_LINK.


If WORKING_DIRECTORY is specified the command will be executed in the directory given. If COMMENT is set, the value will be displayed as a message before the commands are executed at build time. If APPEND is specified the COMMAND and DEPENDS option values are appended to the custom command for the first output specified. There must have already been a previous call to this command with the same output. The COMMENT, WORKING_DIRECTORY, and MAIN_DEPENDENCY options are currently ignored when APPEND is given, but may be used in the future.


If VERBATIM is given then all the arguments to the commands will be passed exactly as specified no matter the build tool used. Note that one level of escapes is still used by the CMake language processor before ADD_CUSTOM_TARGET even sees the arguments. Use of VERBATIM is recommended as it enables correct behavior. When VERBATIM is not given the behavior is platform specific. In the future VERBATIM may be enabled by default. The only reason it is an option is to preserve compatibility with older CMake code.


If the output of the custom command is not actually created as a file on disk it should be marked as SYMBOLIC with SET_SOURCE_FILES_PROPERTIES.


The IMPLICIT_DEPENDS option requests scanning of implicit dependencies of an input file.  The language given specifies the programming language whose corresponding dependency scanner should be used.  Currently only C and CXX language scanners are supported. Dependencies discovered from the scanning are added to those of the custom command at build time.  Note that the IMPLICIT_DEPENDS option is currently supported only for Makefile generators and will be ignored by other generators.


If COMMAND specifies an executable target (created by ADD_EXECUTABLE) it will automatically be replaced by the location of the executable created at build time.  Additionally a target-level dependency will be added so that the executable target will be built before any target using this custom command.  However this does NOT add a file-level dependency that would cause the custom command to re-run whenever the executable is recompiled.


If DEPENDS specifies any target (created by an ADD_* command) a target-level dependency is created to make sure the target is built before any target using this custom command.  Additionally, if the target is an executable or library a file-level dependency is created to cause the custom command to re-run whenever the target is recompiled.


.TP
.B add_custom_target
Add a target with no output so it will always be built.

  add_custom_target(Name [ALL] [command1 [args1...]]
                    [COMMAND command2 [args2...] ...]
                    [DEPENDS depend depend depend ... ]
                    [WORKING_DIRECTORY dir]
                    [COMMENT comment] [VERBATIM])

Adds a target with the given name that executes the given commands. The target has no output file and is ALWAYS CONSIDERED OUT OF DATE even if the commands try to create a file with the name of the target. Use ADD_CUSTOM_COMMAND to generate a file with dependencies. By default nothing depends on the custom target. Use ADD_DEPENDENCIES to add dependencies to or from other targets. If the ALL option is specified it indicates that this target should be added to the default build target so that it will be run every time (the command cannot be called ALL). The command and arguments are optional and if not specified an empty target will be created. If WORKING_DIRECTORY is set, then the command will be run in that directory. If COMMENT is set, the value will be displayed as a message before the commands are executed at build time. Dependencies listed with the DEPENDS argument may reference files and outputs of custom commands created with ADD_CUSTOM_COMMAND.


If VERBATIM is given then all the arguments to the commands will be passed exactly as specified no matter the build tool used. Note that one level of escapes is still used by the CMake language processor before add_custom_target even sees the arguments. Use of VERBATIM is recommended as it enables correct behavior. When VERBATIM is not given the behavior is platform specific. In the future VERBATIM may be enabled by default. The only reason it is an option is to preserve compatibility with older CMake code.

.TP
.B add_definitions
Adds -D define flags to the compilation of source files.

  add_definitions(-DFOO -DBAR ...)

Adds flags to the compiler command line for sources in the current directory and below.  This command can be used to add any flags, but it was originally intended to add preprocessor definitions.  Flags beginning in -D or /D that look like preprocessor definitions are automatically added to the COMPILE_DEFINITIONS property for the current directory.  Definitions with non-trival values may be left in the set of flags instead of being converted for reasons of backwards compatibility.  See documentation of the directory, target, and source file COMPILE_DEFINITIONS properties for details on adding preprocessor definitions to specific scopes and configurations.

.TP
.B add_dependencies
Add a dependency between top-level targets.

  add_dependencies(target-name depend-target1
                   depend-target2 ...)

Make a top-level target depend on other top-level targets.  A top-level target is one created by ADD_EXECUTABLE, ADD_LIBRARY, or ADD_CUSTOM_TARGET.  Adding dependencies with this command can be used to make sure one target is built before another target.  See the DEPENDS option of ADD_CUSTOM_TARGET and ADD_CUSTOM_COMMAND for adding file-level dependencies in custom rules.  See the OBJECT_DEPENDS option in SET_SOURCE_FILES_PROPERTIES to add file-level dependencies to object files.

.TP
.B add_executable
Add an executable to the project using the specified source files.

  add_executable(<name> [WIN32] [MACOSX_BUNDLE] [EXCLUDE_FROM_ALL]
                 source1 source2 ... sourceN)

Adds an executable target called <name> to be built from the source files listed in the command invocation.  The <name> corresponds to the logical target name and must be globally unique within a project.  The actual file name of the executable built is constructed based on conventions of the native platform (such as <name>.exe or just <name>).  


By default the executable file will be created in the build tree directory corresponding to the source tree directory in which the command was invoked.  See documentation of the RUNTIME_OUTPUT_DIRECTORY target property to change this location.  See documentation of the OUTPUT_NAME target property to change the <name> part of the final file name.  


If WIN32 is given the property WIN32_EXECUTABLE will be set on the target created.  See documentation of that target property for details.


If MACOSX_BUNDLE is given the corresponding property will be set on the created target.  See documentation of the MACOSX_BUNDLE target property for details.


If EXCLUDE_FROM_ALL is given the corresponding property will be set on the created target.  See documentation of the EXCLUDE_FROM_ALL target property for details.


The add_executable command can also create IMPORTED executable targets using this signature:


  add_executable(<name> IMPORTED)

An IMPORTED executable target references an executable file located outside the project.  No rules are generated to build it.  The target name has scope in the directory in which it is created and below.  It may be referenced like any target built within the project.  IMPORTED executables are useful for convenient reference from commands like add_custom_command.  Details about the imported executable are specified by setting properties whose names begin in "IMPORTED_".  The most important such property is IMPORTED_LOCATION (and its per-configuration version IMPORTED_LOCATION_<CONFIG>) which specifies the location of the main executable file on disk.  See documentation of the IMPORTED_* properties for more information.

.TP
.B add_library
Add a library to the project using the specified source files.

  add_library(<name> [STATIC | SHARED | MODULE] [EXCLUDE_FROM_ALL]
              source1 source2 ... sourceN)

Adds a library target called <name> to be built from the source files listed in the command invocation.  The <name> corresponds to the logical target name and must be globally unique within a project.  The actual file name of the library built is constructed based on conventions of the native platform (such as lib<name>.a or <name>.lib).


STATIC, SHARED, or MODULE may be given to specify the type of library to be created.  STATIC libraries are archives of object files for use when linking other targets.  SHARED libraries are linked dynamically and loaded at runtime.  MODULE libraries are plugins that are not linked into other targets but may be loaded dynamically at runtime using dlopen-like functionality.  If no type is given explicitly the type is STATIC or SHARED based on whether the current value of the variable BUILD_SHARED_LIBS is true.


By default the library file will be created in the build tree directory corresponding to the source tree directory in which the command was invoked.  See documentation of the ARCHIVE_OUTPUT_DIRECTORY, LIBRARY_OUTPUT_DIRECTORY, and RUNTIME_OUTPUT_DIRECTORY target properties to change this location.  See documentation of the OUTPUT_NAME target property to change the <name> part of the final file name.  


If EXCLUDE_FROM_ALL is given the corresponding property will be set on the created target.  See documentation of the EXCLUDE_FROM_ALL target property for details.


The add_library command can also create IMPORTED library targets using this signature:


  add_library(<name> <SHARED|STATIC|MODULE> IMPORTED)

An IMPORTED library target references a library file located outside the project.  No rules are generated to build it.  The target name has scope in the directory in which it is created and below.  It may be referenced like any target built within the project.  IMPORTED libraries are useful for convenient reference from commands like target_link_libraries.  Details about the imported library are specified by setting properties whose names begin in "IMPORTED_".  The most important such property is IMPORTED_LOCATION (and its per-configuration version IMPORTED_LOCATION_<CONFIG>) which specifies the location of the main library file on disk.  See documentation of the IMPORTED_* properties for more information.

.TP
.B add_subdirectory
Add a subdirectory to the build.

  add_subdirectory(source_dir [binary_dir] 
                   [EXCLUDE_FROM_ALL])

Add a subdirectory to the build. The source_dir specifies the directory in which the source CmakeLists.txt and code files are located. If it is a relative path it will be evaluated with respect to the current directory (the typical usage), but it may also be an absolute path. The binary_dir specifies the directory in which to place the output files. If it is a relative path it will be evaluated with respect to the current output directory, but it may also be an absolute path. If binary_dir is not specified, the value of source_dir, before expanding any relative path, will be used (the typical usage). The CMakeLists.txt file in the specified source directory will be processed immediately by CMake before processing in the current input file continues beyond this command.


If the EXCLUDE_FROM_ALL argument is provided then this subdirectory will not be included in build by default. Users will have to explicitly start a build in the generated output directory. This is useful for having cmake create a build system for a set of examples in a project. One would want cmake to generate a single build system for all the examples, but one may not want the targets to show up in the main build system.

.TP
.B add_test
Add a test to the project with the specified arguments.

  add_test(testname Exename arg1 arg2 ...)

If the ENABLE_TESTING command has been run, this command adds a test target to the current directory. If ENABLE_TESTING has not been run, this command does nothing.  The tests are run by the testing subsystem by executing Exename with the specified arguments.  Exename can be either an executable built by this project or an arbitrary executable on the system (like tclsh).  The test will be run with the current working directory set to the CMakeList.txt files corresponding directory in the binary tree.

.TP
.B aux_source_directory
Find all source files in a directory.

  aux_source_directory(<dir> <variable>)

Collects the names of all the source files in the specified directory and stores the list in the <variable> provided.  This command is intended to be used by projects that use explicit template instantiation.  Template instantiation files can be stored in a "Templates" subdirectory and collected automatically using this command to avoid manually listing all instantiations.


It is tempting to use this command to avoid writing the list of source files for a library or executable target.  While this seems to work, there is no way for CMake to generate a build system that knows when a new source file has been added.  Normally the generated build system knows when it needs to rerun CMake because the CMakeLists.txt file is modified to add a new source.  When the source is just added to the directory without modifying this file, one would have to manually rerun CMake to generate a build system incorporating the new file.

.TP
.B break
Break from an enclosing foreach or while loop.

  break()

Breaks from an enclosing foreach loop or while loop

.TP
.B build_command
Get the command line that will build this project.

  build_command(<variable> <makecommand>)

Sets the given <variable> to a string containing the command that will build this project from the root of the build tree using the build tool given by <makecommand>.  <makecommand> should be msdev, nmake, make or one of the end user build tools.  This is useful for configuring testing systems.

.TP
.B cmake_minimum_required
Set the minimum required version of cmake for a project.

  cmake_minimum_required(VERSION major[.minor[.patch]]
                         [FATAL_ERROR])

If the current version of CMake is lower than that required it will stop processing the project and report an error.  When a version higher than 2.4 is specified the command implicitly invokes


  cmake_policy(VERSION major[.minor[.patch]])

which sets the cmake policy version level to the version specified.  When version 2.4 or lower is given the command implicitly invokes


  cmake_policy(VERSION 2.4)

which enables compatibility features for CMake 2.4 and lower.


The FATAL_ERROR option is accepted but ignored.  It is left from CMake versions 2.4 and lower in which failure to meet the minimum version was a warning by default.

.TP
.B cmake_policy
Manage CMake Policy settings.

As CMake evolves it is sometimes necessary to change existing behavior in order to fix bugs or improve implementations of existing features.  The CMake Policy mechanism is designed to help keep existing projects building as new versions of CMake introduce changes in behavior.  Each new policy (behavioral change) is given an identifier of the form "CMP<NNNN>" where "<NNNN>" is an integer index.  Documentation associated with each policy describes the OLD and NEW behavior and the reason the policy was introduced.  Projects may set each policy to select the desired behavior.  When CMake needs to know which behavior to use it checks for a setting specified by the project.  If no setting is available the OLD behavior is assumed and a warning is produced requesting that the policy be set.


The cmake_policy command is used to set policies to OLD or NEW behavior.  While setting policies individually is supported, we encourage projects to set policies based on CMake versions.


  cmake_policy(VERSION major.minor[.patch])

Specify that the current CMake list file is written for the given version of CMake.  All policies introduced in the specified version or earlier will be set to use NEW behavior.  All policies introduced after the specified version will be reset to use OLD behavior with a warning.  This effectively requests behavior preferred as of a given CMake version and tells newer CMake versions to warn about their new policies.  The policy version specified must be at least 2.4 or the command will report an error.  In order to get compatibility features supporting versions earlier than 2.4 see documentation of policy CMP0001.


  cmake_policy(SET CMP<NNNN> NEW)
  cmake_policy(SET CMP<NNNN> OLD)

Tell CMake to use the OLD or NEW behavior for a given policy.  Projects depending on the old behavior of a given policy may silence a policy warning by setting the policy state to OLD.  Alternatively one may fix the project to work with the new behavior and set the policy state to NEW.


  cmake_policy(PUSH)
  cmake_policy(POP)

Push and pop the current policy setting state on a stack.  Each PUSH must have a matching POP.  This is useful when mixing multiple projects, subprojects, and files included from external projects that may each have been written for a different version of CMake.  Each subdirectory entered by the project automatically pushes a new level on the stack to isolate the subdirectories from their parents.

.TP
.B configure_file
Copy a file to another location and modify its contents.

  configure_file(InputFile OutputFile
                 [COPYONLY] [ESCAPE_QUOTES] [@ONLY])

The Input and Output files have to have full paths.  This command replaces any variables in the input file referenced as ${VAR} or @VAR@ with their values as determined by CMake.  If a variable is not defined, it will be replaced with nothing.  If COPYONLY is specified, then no variable expansion will take place.  If ESCAPE_QUOTES is specified then any substituted quotes will be C-style escaped.  The file will be configured with the current values of CMake variables. If @ONLY is specified, only variables of the form @VAR@ will be replaces and ${VAR} will be ignored.  This is useful for configuring scripts that use ${VAR}. Any occurrences of #cmakedefine VAR will be replaced with either #define VAR or /* #undef VAR */ depending on the setting of VAR in CMake

.TP
.B create_test_sourcelist
Create a test driver and source list for building test programs.

  create_test_sourcelist(sourceListName driverName
                         test1 test2 test3
                         EXTRA_INCLUDE include.h
                         FUNCTION function)

A test driver is a program that links together many small tests into a single executable.  This is useful when building static executables with large libraries to shrink the total required size.  The list of source files needed to build the test driver will be in sourceListName.  DriverName is the name of the test driver program.  The rest of the arguments consist of a list of test source files, can be semicolon separated.  Each test source file should have a function in it that is the same name as the file with no extension (foo.cxx should have int foo(int, char*[]);) DriverName will be able to call each of the tests by name on the command line. If EXTRA_INCLUDE is specified, then the next argument is included into the generated file. If FUNCTION is specified, then the next argument is taken as a function name that is passed a pointer to ac and av.  This can be used to add extra command line processing to each test. The cmake variable CMAKE_TESTDRIVER_BEFORE_TESTMAIN can be set to have code that will be placed directly before calling the test main function.   CMAKE_TESTDRIVER_AFTER_TESTMAIN can be set to have code that will be placed directly after the call to the test main function.

.TP
.B define_property
Define and document custom properties.

  define_property(<GLOBAL | DIRECTORY | TARGET | SOURCE |
                   TEST | VARIABLE | CACHED_VARIABLE>
                   PROPERTY <name> [INHERITED]
                   BRIEF_DOCS <brief-doc>
                   FULL_DOCS <full-doc>)

Define one property in a scope for use with the set_property and get_property commands.  This is primarily useful to associate documentation with property names that may be retrieved with the get_property command.  The first argument determines the kind of scope in which the property should be used.  It must be one of the following:


  GLOBAL    = associated with the global namespace
  DIRECTORY = associated with one directory
  TARGET    = associated with one target
  SOURCE    = associated with one source file
  TEST      = associated with a test named with add_test command
  VARIABLE  = documents a CMake language variable
  CACHED_VARIABLE = documents a CMake cache variable

Note that unlike set_property and get_property no actual scope needs to be given; only the kind of scope is important.


The required PROPERTY option is immediately followed by the name of the property being defined.


If the INHERITED option then the get_property command will chain up to the next higher scope when the requested property is not set in the scope given to the command.  DIRECTORY scope chains to GLOBAL.  TARGET, SOURCE, and TEST chain to DIRECTORY.


The BRIEF_DOCS and FULL_DOCS options are followed by strings to be associated with the property as its brief and full documentation.  Corresponding options to the get_property command will retrieve the documentation.

.TP
.B else
Starts the else portion of an if block.

  else(expression)

See the if command.

.TP
.B elseif
Starts the elseif portion of an if block.

  elseif(expression)

See the if command.

.TP
.B enable_language
Enable a language (CXX/C/Fortran/etc)

  enable_language(languageName [OPTIONAL] )

This command enables support for the named language in CMake. This is the same as the project command but does not create any of the extra variables that are created by the project command. Example languages are CXX, C, Fortran.


If OPTIONAL is used, use the CMAKE_<languageName>_COMPILER_WORKS variable to check whether the language has been enabled successfully.

.TP
.B enable_testing
Enable testing for current directory and below.

  enable_testing()

Enables testing for this directory and below.  See also the add_test command.  Note that ctest expects to find a test file in the build directory root.  Therefore, this command should be in the source directory root.

.TP
.B endforeach
Ends a list of commands in a FOREACH block.

  endforeach(expression)

See the FOREACH command.

.TP
.B endfunction

QMOLEDEV/cmake-2.6.0/Source/CursesDialog/ccmake.cxx
/*=========================================================================

  Program:   CMake - Cross-Platform Makefile Generator
  Module:    $RCSfile: ccmake.cxx,v $
  Language:  C++
  Date:      $Date: 2007-12-13 22:56:50 $
  Version:   $Revision: 1.37 $

  Copyright (c) 2002 Kitware, Inc., Insight Consortium.  All rights reserved.
  See Copyright.txt or http://www.cmake.org/HTML/Copyright.html for details.

     This software is distributed WITHOUT ANY WARRANTY; without even 
     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
     PURPOSE.  See the above copyright notices for more information.

=========================================================================*/
#include "../cmCacheManager.h"
#include "../cmSystemTools.h"
#include "../cmake.h"
#include "../cmDocumentation.h"

#include <signal.h>
#include <sys/ioctl.h>

#include "cmCursesMainForm.h"
#include "cmCursesStandardIncludes.h"

#include <form.h>

//----------------------------------------------------------------------------
static const char * cmDocumentationName[][3] =
{
  {0,
   "  ccmake - Curses Interface for CMake.", 0},
  {0,0,0}
};

//----------------------------------------------------------------------------
static const char * cmDocumentationUsage[][3] =
{
  {0,
   "  ccmake <path-to-source>\n"
   "  ccmake <path-to-existing-build>", 0},
  {0,0,0}
};

//----------------------------------------------------------------------------
static const char * cmDocumentationDescription[][3] =
{
  {0,
   "The \"ccmake\" executable is the CMake curses interface.  Project "
   "configuration settings may be specified interactively through "
   "this GUI.  Brief instructions are provided at the bottom of the "
   "terminal when the program is running.", 0},
  CMAKE_STANDARD_INTRODUCTION,
  {0,0,0}
};

//----------------------------------------------------------------------------
static const char * cmDocumentationOptions[][3] =
{
  CMAKE_STANDARD_OPTIONS_TABLE,
  {0,0,0}
};

//----------------------------------------------------------------------------
static const char * cmDocumentationSeeAlso[][3] =
{
  {0, "cmake", 0},
  {0, "ctest", 0},
  {0, 0, 0}
};

cmCursesForm* cmCursesForm::CurrentForm=0;

extern "C"
{

void onsig(int)
{
  if (cmCursesForm::CurrentForm)
    {
    endwin();
    initscr(); /* Initialization */ 
    noecho(); /* Echo off */ 
    cbreak(); /* nl- or cr not needed */ 
    keypad(stdscr,TRUE); /* Use key symbols as 
                            KEY_DOWN*/ 
    refresh();
    int x,y;
    getmaxyx(stdscr, y, x);
    cmCursesForm::CurrentForm->Render(1,1,x,y);
    cmCursesForm::CurrentForm->UpdateStatusBar();
    }
  signal(SIGWINCH, onsig);
}
 
}

void CMakeErrorHandler(const char* message, const char* title, bool&, void* clientData)
{
  cmCursesForm* self = static_cast<cmCursesForm*>( clientData );
  self->AddError(message, title);
}

int main(int argc, char** argv)
{
  cmSystemTools::FindExecutableDirectory(argv[0]);
  cmDocumentation doc;
  if(doc.CheckOptions(argc, argv))
    {
    cmake hcm;
    std::vector<cmDocumentationEntry> commands;
    std::vector<cmDocumentationEntry> compatCommands;
    std::vector<cmDocumentationEntry> generators;
    hcm.GetCommandDocumentation(commands, true, false);
    hcm.GetCommandDocumentation(compatCommands, false, true);
    hcm.GetGeneratorDocumentation(generators);
    doc.SetName("ccmake");
    doc.SetSection("Name",cmDocumentationName);
    doc.SetSection("Usage",cmDocumentationUsage);
    doc.SetSection("Description",cmDocumentationDescription);
    doc.SetSection("Generators",generators);
    doc.SetSection("Options",cmDocumentationOptions);
    doc.SetSection("Command",commands);
    doc.SetSection("Compatibility Commands",compatCommands);
    doc.SetSeeAlsoList(cmDocumentationSeeAlso);
    return doc.PrintRequestedDocumentation(std::cout)? 0:1;
    }  
  
  bool debug = false;
  unsigned int i;
  int j;
  std::vector<std::string> args;
  for(j =0; j < argc; ++j)
    {
    if(strcmp(argv[j], "-debug") == 0)
      {
      debug = true;
      }
    else
      {
      args.push_back(argv[j]);
      }
    }

  std::string cacheDir = cmSystemTools::GetCurrentWorkingDirectory();
  for(i=1; i < args.size(); ++i)
    {
    std::string arg = args[i];
    if(arg.find("-B",0) == 0)
      {
      cacheDir = arg.substr(2);
      }
    }

  cmSystemTools::DisableRunCommandOutput();

  if (debug)
    {
    cmCursesForm::DebugStart();
    }

  initscr(); /* Initialization */ 
  noecho(); /* Echo off */ 
  cbreak(); /* nl- or cr not needed */ 
  keypad(stdscr,TRUE); /* Use key symbols as 
                          KEY_DOWN*/ 

  signal(SIGWINCH, onsig);

  int x,y;
  getmaxyx(stdscr, y, x);
  if ( x < cmCursesMainForm::MIN_WIDTH  || 
       y < cmCursesMainForm::MIN_HEIGHT )
    {
    endwin();
    std::cerr << "Window is too small. A size of at least "
              << cmCursesMainForm::MIN_WIDTH << " x " 
              <<  cmCursesMainForm::MIN_HEIGHT
              << " is required to run ccmake." <<  std::endl;
    return 1;
    }


  cmCursesMainForm* myform;

  myform = new cmCursesMainForm(args, x);
  if(myform->LoadCache(cacheDir.c_str()))
    {
    curses_clear();
    touchwin(stdscr);
    endwin();
    delete myform;
    std::cerr << "Error running cmake::LoadCache().  Aborting.\n";
    return 1;
    }

  cmSystemTools::SetErrorCallback(CMakeErrorHandler, myform);

  cmCursesForm::CurrentForm = myform;

  myform->InitializeUI();
  if ( myform->Configure(1) == 0 )
    {
    myform->Render(1, 1, x, y);
    myform->HandleInput();
    }
  
  // Need to clean-up better
  curses_clear();
  touchwin(stdscr);
  endwin();
  delete cmCursesForm::CurrentForm;
  cmCursesForm::CurrentForm = 0;

  std::cout << std::endl << std::endl;
  
  return 0;

}
